### PR TITLE
Skip stackmaps for yk_trace_basicblock.

### PR DESCRIPTION
**Pull request by Pavel-Durov**
Related: https://github.com/ykjit/yk/pull/939
Source: https://github.com/ykjit/yk/pull/939#issuecomment-1909980698

The yk_trace_basicblock call doesn't require stackmaps therefore we can skip stackmap adjustment for those functions.

Tests:
```
Testing Time: 405.04s
  Skipped          :    46
  Unsupported      :   885
  Passed           : 92778
  Expectedly Failed:   211
```

### CODE REVIEW

**Comment by ltratt** [comment]
Don't we need to guard this to only when we're compiling in swt mode? Also, what performance difference does this make (roughly speaking)?

**Comment by Pavel-Durov** [comment]
> Don't we need to guard this to only when we're compiling in swt mode? 

Yes, I'll add the guard.

> Also, what performance difference does this make (roughly speaking)?

Performance difference between `swt` and `hwt`?

File: llvm/lib/CodeGen/Yk/FixStackmapsSpillReloads.cpp [hunk 1]
```diff
@@ -103,6 +104,13 @@ bool FixStackmapsSpillReloads::runOnMachineFunction(MachineFunction &MF) {
         // removed during lowering?
         if (MI.getOpcode() != TargetOpcode::STACKMAP &&
             MI.getOpcode() != TargetOpcode::PATCHPOINT) {
+          MachineOperand Op = MI.getOperand(0);
+          if (Op.isGlobal() && Op.getGlobal()->getGlobalIdentifier() == YK_TRACE_FUNCTION) {
```
Comment (Pavel-Durov) [inline llvm/lib/CodeGen/Yk/FixStackmapsSpillReloads.cpp:108]: Add a guard here so it will only take affect in `swt` mode

**Comment by ltratt** [comment]
>>    Also, what performance difference does this make (roughly speaking)?
> Performance difference between swt and hwt?

The performance difference of not adding stackmaps to yk_trace_basicblock vs. adding it.

**Comment by Pavel-Durov** [comment]
> Don't we need to guard this to only when we're compiling in swt mode? Also, what performance difference does this make (roughly speaking)?

@ltratt LLVM passes are compile time so the impact of adding a guard here will only affect compilation time which is insignificant for the runtime performance impact.

**Comment by Pavel-Durov** [comment]
## Benchmarks
### With yk_trace_basicblock stackmaps
```
hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/closure.lua' 

Benchmark 1: ./src/lua ./tests/closure.lua
  Time (mean ± σ):      4.439 s ±  0.411 s    [User: 4.569 s, System: 0.186 s]
  Range (min … max):    3.597 s …  5.049 s    30 runs

hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/strings.lua' 
Benchmark 1: ./src/lua ./tests/strings.lua
  Time (mean ± σ):      3.481 s ±  0.351 s    [User: 3.521 s, System: 0.097 s]
  Range (min … max):    3.245 s …  4.120 s    30 runs
```
### Without yk_trace_basicblock stackmaps

```
hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/closure.lua'
Benchmark 1: ./src/lua ./tests/closure.lua
  Time (mean ± σ):      4.022 s ±  0.464 s    [User: 4.020 s, System: 0.127 s]
  Range (min … max):    3.318 s …  4.736 s    30 runs

hyphyperfine --min-runs 30 './src/lua ./tests/strings.lua' 
Benchmark 1: ./src/lua ./tests/strings.lua
  Time (mean ± σ):      3.506 s ±  0.441 s    [User: 3.537 s, System: 0.096 s]
  Range (min … max):    2.649 s …  4.116 s    30 runs
```

**Comment by ltratt** [comment]
We're still getting seriously noisy numbers here, so it's very hard to conclude much :/ The range for `strings.lua` is "2.649s …  4.116s" which I'm really struggling to come up with an explanation for.

**Comment by Pavel-Durov** [comment]
Ran again multiple times on a completely quite machine (no anomalies were reported by hyperfine)
The variation is not as big as before for `strings.lua`.

## Benchmarks

### With yk_trace_basicblock stackmaps
```
hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/closure.lua' 
Benchmark 1: ./src/lua ./tests/closure.lua
  Time (mean ± σ):      4.112 s ±  0.097 s    [User: 4.139 s, System: 0.064 s]
  Range (min … max):    3.794 s …  4.325 s    30 runs

hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/strings.lua' 
Benchmark 1: ./src/lua ./tests/strings.lua
  Time (mean ± σ):      3.896 s ±  0.098 s    [User: 3.937 s, System: 0.051 s]
  Range (min … max):    3.711 s …  4.069 s    30 runs
```
### Without yk_trace_basicblock stackmaps

```
hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/closure.lua' 
Benchmark 1: ./src/lua ./tests/closure.lua
  Time (mean ± σ):      3.503 s ±  0.069 s    [User: 3.520 s, System: 0.060 s]
  Range (min … max):    3.317 s …  3.624 s    30 runs
 
hyperfine --warmup 10 --min-runs 30 './src/lua ./tests/strings.lua' 
Benchmark 1: ./src/lua ./tests/strings.lua
  Time (mean ± σ):      3.248 s ±  0.070 s    [User: 3.280 s, System: 0.045 s]
  Range (min … max):    3.112 s …  3.371 s    30 runs
 ```

**Comment by ltratt** [comment]
Ok those look better! So roughly a 10% difference from removing stackmaps here. Very interesting -- thank you for this!

[If you're wondering why I was wondering, that means I can now guesstimate the cost of stackmaps on performance of the rest of the interpreter as ~5-7%. Not great, but not quite as bad as I feared.]

**Comment by ltratt** [comment]
@ptersilie I'm fine with this PR if you are!

**Comment by ptersilie** [comment]
Looks good.

**Pull request merged**